hack/tools/cmd/gomodcheck/main.go

```diff
@@ -0,0 +1,203 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"go.uber.org/zap"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	modFile = "./go.mod"
+)
+
+type config struct {
+	UpstreamRefs    []string `yaml:"upstreamRefs"`
+	ExcludedModules []string `yaml:"excludedModules"`
+}
+
+type upstream struct {
+	Ref     string `json:"ref"`
+	Version string `json:"version"`
+}
+
+// representation of an out of sync module
+type oosMod struct {
+	Name      string     `json:"name"`
+	Version   string     `json:"version"`
+	Upstreams []upstream `json:"upstreams"`
+}
+
+func main() {
+	l, _ := zap.NewProduction()
+	logger := l.Sugar()
+
+	if len(os.Args) < 2 {
+		fmt.Printf("USAGE: %s [PATH_TO_CONFIG_FILE]\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	// --- 1. parse config
+	b, err := os.ReadFile(os.Args[1])
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	cfg := new(config)
+	if err := yaml.Unmarshal(b, cfg); err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	excludedMods := make(map[string]any)
+	for _, mod := range cfg.ExcludedModules {
+		excludedMods[mod] = nil
+	}
+
+	// --- 2. project mods
+	deps, err := parseModFile()
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	// --- 3. upstream mods (holding upstream refs)
+	upstreamModGraph, err := getUpstreamModGraph(cfg.UpstreamRefs)
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	oosMods := make([]oosMod, 0)
+
+	// --- 4. validate
+	// for each module in our project,
+	// if it matches an upstream module,
+	// then for each upstream module,
+	// if project module version doesn't match upstream version,
+	// then we add the version and the ref to the list of out of sync modules.
+	for mod, version := range deps {
+		if _, ok := excludedMods[mod]; ok {
+			logger.Infof("skipped excluded module: %s", mod)
+			continue
+		}
+
+		if versionToRef, ok := upstreamModGraph[mod]; ok {
+			upstreams := make([]upstream, 0)
+
+			for upstreamVersion, upstreamRef := range versionToRef {
+				if version != upstreamVersion {
+					upstreams = append(upstreams, upstream{
+						Ref:     upstreamRef,
+						Version: upstreamVersion,
+					})
+				}
+			}
+
+			if len(upstreams) > 0 {
+				oosMods = append(oosMods, oosMod{
+					Name:      mod,
+					Version:   version,
+					Upstreams: upstreams,
+				})
+			}
+		}
+	}
+
+	if len(oosMods) == 0 {
+		fmt.Println("Success! 🎉")
+		os.Exit(0)
+	}
+
+	b, err = json.MarshalIndent(map[string]any{"outOfSyncModules": oosMods}, "", "  ")
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(string(b))
+	os.Exit(1)
+}
+
+var (
+	cleanMods     = regexp.MustCompile(`\t| *//.*`)
+	modDelimStart = regexp.MustCompile(`^require.*`)
+	modDelimEnd   = ")"
+)
+
+func parseModFile() (map[string]string, error) {
+	b, err := os.ReadFile(modFile)
```

Comment:
WDYT about using this package to parse the mod file? (https://pkg.go.dev/golang.org/x/mod/modfile)

(not sure how easy it is to use)

Just want to avoid that this code here breaks when they change the format in some way

Reply:
That is an elegant alternative. I'll use it instead, thanks.

NB: I was initially looking into `modload.LoadModGraph` but is an internal package: https://github.com/golang/go/blob/master/src/cmd/go/internal/modload/buildlist.go#L565

Reply:
Fixed in 6449beb3.

Reply:
Nice!